src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -0,0 +1,320 @@
+        internal static DiagnosticDescriptor ImplementsEmptyPreviewInterfaceRule = new(RuleId,
+                                                                                      s_localizableTitle,
+                                                                                      s_implementsEmptyPreviewInterface,
+                                                                                      DiagnosticCategory.Usage,
+                                                                                      DiagnosticSeverity.Error,
+                                                                                      isEnabledByDefault: true,
+                                                                                      s_localizableDescription,
+                                                                                      helpLinkUri: @" https://aka.ms/dotnet-warnings/preview-features");
+
+#pragma warning restore RS0030
+        private const string RequiresPreviewFeaturesAttribute = nameof(RequiresPreviewFeaturesAttribute);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralPreviewFeatureAttributeRule, StaticAbstractIsPreviewFeatureRule, ImplementsEmptyPreviewInterfaceRule);
+
+        private enum PreviewFeatureUsageType
+        {
+            General = 0,
+            StaticAbstract,
+            EmptyInterface
+        }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(context =>
+            {
+                Compilation compilation = context.Compilation;
+                IAssemblySymbol? thisAssembly = compilation.Assembly;
+                ImmutableArray<AttributeData> attributes = thisAssembly.GetAttributes();
+
+                AttributeData? attribute = attributes.FirstOrDefault(x => x.AttributeClass.Name == RequiresPreviewFeaturesAttribute);
+                if (!(attribute is null))
+                {
+                    // This assembly has enabled preview attributes.
+                    return;
+                }
+
+                ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols = new ConcurrentDictionary<ISymbol, bool>();
+                ConcurrentDictionary<ISymbol, PreviewFeatureUsageType> requiresPreviewFeaturesSymbolsToUsageType = new ConcurrentDictionary<ISymbol, PreviewFeatureUsageType>();
+
+                if (context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemRuntimeCompilerServicesRuntimeFeature, out INamedTypeSymbol? runtimeFeatureType))
+                {
+                    if (runtimeFeatureType != null)
+                    {
+                        HasPreviewAttribute(runtimeFeatureType, requiresPreviewFeaturesSymbols);
```

Comment:
it's a bit strange, need to check the type loaded from the runtime?  not existing in the user code?

Reply:
Yup, because static abstract is considered preview if runtimeFeatureType has the Preview attribute on it. So, in a future release, if runtimeFeatureType is not marked as preview, we won't consider static abstract as preview.

Reply:
Fixed now. I'm looking for the attribute on RuntimeFeature.VirtualStaticsInInterfaces now